### PR TITLE
Contracts listener

### PR DESCRIPTION
**Pull request by MixailE**
https://trello.com/c/eu8fUX1F/743-sc-add-listeners-to-contractsstore-to-keep-it-updated

### CODE REVIEW

File: src/stores/OriginalContractsStore.ts [hunk 1]
```diff
@@ -1,24 +1,61 @@
...
+      OriginalContractsStore.locked = true
+      OriginalContractsStore.contractsOwnedState = await getOwnedERC721(
+        OriginalContractsStore.contractsOwnedStateAccount,
+        OriginalContractsStore.lastBlockId
+          ? OriginalContractsStore.lastBlockId + 1
+          : 0,
+        currentBlockId,
+        currentState
+      )
+      OriginalContractsStore.locked = false
```
Comment (backmeupplz) [inline src/stores/OriginalContractsStore.ts:44]: Should this happen last?

File: src/stores/OriginalContractsStore.ts [hunk 2]
```diff
@@ -1,24 +1,61 @@
...
+import defaultProvider from 'helpers/defaultProvider'
 import getOwnedERC721 from 'helpers/getOwnedERC721'
 
 interface ContractsStoreType {
-  contractsOwned: Promise<string[]>
+  contractsOwnedState: { [token: string]: number }
+  contractsOwnedStateAccount?: string
+  contractsOwned: string[]
+  lastBlockId?: number
+  locked: boolean
```
Comment (backmeupplz) [inline src/stores/OriginalContractsStore.ts:12]: Can we move the util state (e.g. `locked`) to a separate global variable? Don't want it to be a `proxy`

File: src/stores/OriginalContractsStore.ts [hunk 3]
```diff
@@ -1,24 +1,61 @@
...
+      OriginalContractsStore.contractsOwnedState = {}
+      OriginalContractsStore.contractsOwned = []
+    }
+
+    if (!OriginalContractsStore.contractsOwnedStateAccount) return
+
+    if (!OriginalContractsStore.locked) {
+      const currentBlockId = await defaultProvider.getBlockNumber()
+      const currentState = await OriginalContractsStore.contractsOwnedState
+      OriginalContractsStore.locked = true
```
Comment (backmeupplz) [inline src/stores/OriginalContractsStore.ts:35]: Should this happen first?

File: src/stores/OriginalContractsStore.ts (hunk 1 above)
Comment (backmeupplz) [reply src/stores/OriginalContractsStore.ts:44]: What if something throws an exception? Should we do `try/catch/finally`?

File: src/stores/OriginalContractsStore.ts [hunk 4]
```diff
@@ -1,24 +1,61 @@
...
 const OriginalContractsStore = proxy<ContractsStoreType>({
-  contractsOwned: Promise.resolve([]),
-  fetchContractsOwned() {
-    OriginalContractsStore.contractsOwned = WalletStore.account
-      ? getOwnedERC721(WalletStore.account)
-      : Promise.resolve([])
+  locked: false,
+  contractsOwnedState: {},
+  contractsOwned: [],
+  async fetchContractsOwned() {
```
Comment (backmeupplz) [inline src/stores/OriginalContractsStore.ts:20]: Maybe `fetchMoreContractsOwned`?

File: src/stores/OriginalContractsStore.ts [hunk 5]
```diff
@@ -1,24 +1,61 @@
...
+    if (
+      OriginalContractsStore.contractsOwnedStateAccount !== WalletStore.account
```
Comment (backmeupplz) [inline src/stores/OriginalContractsStore.ts:22]: Should this happen in the `account` listener?

**Review by backmeupplz — CHANGES_REQUESTED** [review]

File: src/stores/OriginalContractsStore.ts (hunk 5 above)
Comment (MixailE) [reply src/stores/OriginalContractsStore.ts]: Maybe I misunderstood, but checking the selected account should include both listeners

File: src/stores/OriginalContractsStore.ts [hunk 6]
```diff
@@ -1,24 +1,41 @@
...
+    ].getOwnedERC721()
   },
 })
 
 subscribeKey(WalletStore, 'account', () => {
-  OriginalContractsStore.fetchContractsOwned()
+  OriginalContractsStore.fetchMoreContractsOwned()
+})
+
+defaultProvider.on('block', () => {
```
Comment (T-Damer) [inline src/stores/OriginalContractsStore.ts:37]: Do we need to do `() => {...}` here? Maybe we can get rid of `{}`

**Review by T-Damer — APPROVED** [review]
lgtm

File: src/stores/OriginalContractsStore.ts [hunk 7]
```diff
@@ -1,24 +1,41 @@
...
+
+    OriginalContractsStore.contractsOwned = await connectedAccounts[
+      WalletStore.account
+    ].getOwnedERC721()
   },
 })
 
 subscribeKey(WalletStore, 'account', () => {
-  OriginalContractsStore.fetchContractsOwned()
+  OriginalContractsStore.fetchMoreContractsOwned()
```
Comment (backmeupplz) [inline src/stores/OriginalContractsStore.ts:34]: Same here, let's do `subscribeKey(WalletStore, 'account', OriginalContractsStore.fetchMoreContractsOwned)`

**Review by backmeupplz — CHANGES_REQUESTED** [review]

**Review by backmeupplz — APPROVED** [review]

**Pull request merged**